File: Python3/1647. Minimum Deletions to Make Character Frequencies Unique.py
class Solution:
    def minDeletions(self, s: str) -> int:
        chars = Counter(s)
        chars = sorted(chars.items(), key = lambda x: x[1])
        ans = 0
        #print(chars)
        count = ['0'] * (chars[-1][1] + 1)
        for c, n in chars:
            # this number has not been claimed
            if count[n] == '0':
                count[n] = c
                continue
            
            i = n
            # search for an unclaimed position
            while count[i] != '0':
                ans += 1
                i -= 1
            
            # exclude 0 count of this char
            if i != 0:
                count[i] = c
        
        return ans
            

File: Java/1647. Minimum Deletions to Make Character Frequencies Unique.java
class Solution {
    public int minDeletions(String s) {
        int[] chars = new int[26];
        for (char c: s.toCharArray()) {
            chars[c - 'a']++;
        }
        Arrays.sort(chars);
        
        int ans = 0, allowed = s.length();
        for (int i = 25; i >= 0; i--) {
            if (chars[i] > allowed) {
                ans += chars[i] - allowed;
                chars[i] = allowed;
            }
            allowed = Math.max(0, chars[i] - 1);
        }
        
        return ans;
    }
}